add item api

sugoiii opened the pull request:
- Added Item Serializer
- Implemented features from API Spec
  - GET item list, retrieve item
  - POST item
  - PATCH item update, close

sugoiii committed:
Current Status

- 기본적인 serializer을 구현한 상태입니다.
- item을 create할 때 optioncategory와 optionitems를 함께 create하는
방법을 찾는 중입니다.
- model의 경우 model update 머지 후 다시 수정해주어야 합니다.

sugoiii committed:
Basic implementation for Item_Collection

- limit, offset param에 대해서만 현재 get 가능
- options를 create할수가 없다.....

samjo-nyang committed:
add payment api (done #64)

- add info field to the payment model
- add some serializers - option_category, option_item, payment and record
- add payment list retrive / join / update / delete logic
- some test will be needed

samjo-nyang committed:
modify transition map

- BANNED -> JOINED: host
- PENDING -> BANNED: host
- PENDING -> JOINED: host
- JOINED -> BANNED: host
- JOINED -> PAID: participant
- DISPUTED -> PAID: participant
- DISPUTED -> JOINED: participant
- DISPUTED -> CONFIRMED: host
- PAID -> JOINED: participant
- PAID -> DISPUTED: host
- PAID -> CONFIRMED: host

sugoiii committed:
Merge branch 'issue-51_item_API' of https://github.com/sparcs-kaist/sparcs09 into issue-51_item_API

sugoiii committed:
Merge branch 'develop' into issue-51_item_API

sugoiii committed:
implement item

- Implemented Item, Item-Collection
- Loggings are not implemented Now

sugoiii committed:
Update Item

- Updated filtering of Payments
- Changed Item's User Serializer to UserPublicSerializer

sugoiii committed:
Ban Item Full Update

sugoiii committed:
Updated Item Update Logic

samjo-nyang commented on sparcs09-api/apps/core/admin.py:
```diff
@@ -6,7 +6,7 @@
 class ItemAdmin(admin.ModelAdmin):
     list_display = ('title', 'host', 'price', 'join_type', 'created_date',
-                    'deadline', 'delivery_date')
+                    'deadline', 'delivery_date', 'pk')
```
pk는 굳이 admin에 존재할 필요가 없어보입니다. 그리고 id 아닌지 확인부탁드립니다.

samjo-nyang commented on sparcs09-api/apps/core/serializers.py:
```diff
@@ -0,0 +1,89 @@
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+
+from apps.core.models import Comment, Content, Item, OptionCategory, \
```
Python에서 \을 사용해서 강제로 라인을 나누는것은 권장되지 않습니다. [나 (이 있으면 개행해도 정상적으로 인식되므로, 아래와 같이 작성해주세요. https://www.python.org/dev/peps/pep-0008/#maximum-line-length
```python
import (
    AAA, BBB, CCC,
    DDD, EEE, FFF,
)
```

samjo-nyang commented on sparcs09-api/apps/core/serializers.py:
```diff
@@ -0,0 +1,89 @@
+from django.utils import timezone
```
serializers.py의 단일 파일로 작성하지 마시고 대상 객체에 따라 serializers 폴더 안에 파일로 만들어 쪼개주세요! 관리하기 편해집니다

samjo-nyang commented on sparcs09-api/apps/core/serializers.py:
```diff
@@ -0,0 +1,89 @@
+class ContentSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Content
+        # TODO - Change type to kind
```
확인해주세요

samjo-nyang commented on sparcs09-api/apps/core/serializers.py:
```diff
@@ -0,0 +1,89 @@
+                'deadline': 'Deadline should be later than today'
+            })
+
+        return super(ItemSerializer, self).to_internal_value(data)
```
Python3에서는 `super().xxx`로 써도 됩니다. 바꿔주세요><

samjo-nyang commented on sparcs09-api/apps/core/serializers.py:
```diff
@@ -0,0 +1,89 @@
+        if not is_aware(deadline):
+            deadline = make_aware(deadline)
+
+        if int(price) < 100:
```
validation은 따로 validate나 validate_xxx 메서드에 작성부탁드립니다.

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+from django.utils import timezone
```
지금보니 파일에 중복이 있네요. `serializers.py` 파일을 사용하지 않는다면 삭제해주세요

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+
+    def create(self, validated_data):
+        categories = validated_data.pop('option_categories')
+        print(categories)
```
debug print는 pr 날리기 전 삭제해주세요

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+
+        return item
+
+    def update(self, instance, validated_data):
```
`ItemCreateSerializer`인데 update logic이 같이 있네요. `ItemUpdateSerializer` 하나 만들어서 분리하는게 좋을 것 같습니다.

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+        price = data.get('price', self.instance.price)
+        deadline = data.get('deadline', self.instance.deadline)
+        if not is_aware(deadline):
+            deadline = make_aware(deadline)
```
deadline 변수를 만들어놓고, 사용하지 않았습니다. 또한, validate 메서드는 side-effect free였으면 좋겠습니다. timezone aware datetime field가 필요하다면 https://stackoverflow.com/questions/17331578/django-rest-framework-timezone-aware-renderers-parsers 이걸 참고하셔서 구현해주세요

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+        if not is_aware(deadline):
+            deadline = make_aware(deadline)
+
+        if int(price) < 100:
```
이미 drf에서는 integer field에 min값을 지원하는 것으로 알고있습니다. 확인부탁드려요

samjo-nyang commented on sparcs09-api/apps/core/urls.py:
```diff
@@ -0,0 +1,24 @@
+"""sparcs09 URL Configuration
```
모든 routing은 `router.py`에 의해 이루어지는 것이 통일성에 좋을 것 같습니다. 삭제부탁드립니다.

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+        return Response(serializer.data)
+
+    def filter_queryset(self, request):
+        limit = int(request.query_params.get('limit', 10))
```
limit과 offset을 처리하는 코드는 이미 구현되어있습니다. `apps.core.utils.get_limit_offset(request.GET)` 참고하시면 됩니다.

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+    def filter_queryset(self, request):
+        limit = int(request.query_params.get('limit', 10))
+        offset = int(request.query_params.get('offset', 0))
+        status = int(request.query_params.get('status', 0))
```
status가 빈 값이 아니나, 숫자로 변환될 수 없는 값인경우 오류가 발생합니다. 마찬가지로 `apps.core.utils.to_int` 메서드를 사용하시면 됩니다.

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+        status = int(request.query_params.get('status', 0))
+        sid = request.query_params.get('sid', '')
+
+        if status > 63 or status < 0:
```
아래 주석에서는 128까지 허용되는데 여기서는 63까지로 제한했네요. 확인 부탁드립니당

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+        if status > 63 or status < 0:
+            raise ParseError("Wrong status number")
+
+        if status & 0b111101 != 0:
```
`if status & 0b111101 != 0 and sid == ''` 처럼 한줄에 쓸 수 있습니다. 그리고 단순한 로직이 아닌만큼 저 0b111101이 나타내는 값이 무엇인지 주석부탁드립니다.

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+        sid = request.query_params.get('sid', '')
+
+        if status > 63 or status < 0:
+            raise ParseError("Wrong status number")
```
코드 전체에서 문자열을 감싸는 따옴표를 통일할 필요가 있습니다. " 대신 '를 사용해주세요

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+            item_ids = (Payment.objects.filter(participant__username=sid)
+                        .values_list('item', flat=True)
+                        )
+            query_1 = Item.objects.exclude(id__in=item_ids)
```
chaining해서 `query_1 = xxxxx.exclude(...).filter(...)` 로 쓸 수 있습니다.

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+                        )
+            query_1 = Item.objects.exclude(id__in=item_ids)
+            query_1 = query_1.filter(join_type=0)
+            queryset = queryset | query_1
```
`queryset |= query_1`로 쓰는 것이 더 간단해보입니다.

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
     # [POST] /items/
     def create(self, request):
-        pass
+        data = request.data.copy()
```
굳이 copy할 필요가 있나요?

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
+        if not serializer.is_valid():
+            return Response(serializer.errors,
+                            status=status.HTTP_400_BAD_REQUEST)
+        else:
```
굳이 else를 쓰기보다는 그냥 바로 `serializer.save()`를 써도 될 것 같습니다. 로깅 부탁드립니다

samjo-nyang replied:
수정부탁드립니당

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
     # [GET] /items/<pk>/
     def retrieve(self, request, pk=None):
-        pass
+        queryset = Item.objects.all()
```
`self.get_object()` 메서드를 사용해서 106 ~ 107줄과 동일한 일을 할 수 있습니다.

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
     def update(self, request, pk=None):
-        pass
+        return Response({
+            'detail': 'Updating Item is not allowed.',
```
대소문자 주의해주세요><

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -27,27 +30,123 @@ class ItemViewSet(viewsets.ModelViewSet):
     # [DELETE] /items/<pk>/
     def destroy(self, request, pk=None):
-        pass
+        item = self.get_object()
```
참여한 사용자가 있다면 삭제를 막아야 합니다. 구현 부탁드립니다

samjo-nyang reviewed (changes requested):
일단 실행은 시켜보지 않았고 눈에 띄는것만 리뷰했습니다.  확인부탁드립니다~

sugoiii committed:
Fixed issues from pull request

- Fixed issues from first pull request
  - Implemented ItemUpdateSerializer
  - Added logger
    - Item create, update, delete, close
  - Plus lots of minor fixes....

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+                'deadline': 'Deadline should be later than now'
+            })
+
+        if not (data.get('description') or data.get('thumbnail')):
```
이거 description이랑 thumbnail 없을 때 오류 안뜨나요?

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+                  'join_type', 'deadline', 'delivery_date',
+                  'option_categories']
+
+    # def validate(self, data):
```
의미 없는 주석 삭제부탁드립니당

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+        # read_only_fields = ['host', 'created_date', 'is_deleted']
+        write_free_fields = ['deadline', 'delivery_date']
+
+        payment_occured = len(Payment.objects.filter(item=instance.id)) != 0
```
`len(xxx) != 0` 보다는 `xxx.count() != 0`을 추천합니다

samjo-nyang replied:
아 죄송합니다 `xxx.exists()`가 가장 좋을 것 같습니다

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+        instance.save()
+
+        if categories is not None:
+            if payment_occured:
```
한 if로 쓸 수 있습니다

samjo-nyang commented on sparcs09-api/apps/core/serializers/item.py:
```diff
@@ -1,9 +1,122 @@
+                item=instance.id
+                )
+
+            for category in current_categories:
```
OptionCategory.objects.filter(...).delete()로 충분합니다

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -1,53 +1,189 @@
 class ItemViewSet(viewsets.ModelViewSet):
     lookup_value_regex = '\d+'
-    queryset = Item.objects.all()
+    queryset = None
```
queryset을 그대로 놓고 get_queryset logic은 view logic에 넣는게 좋아보입니다...

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -1,53 +1,189 @@
+        if not serializer.is_valid():
+            return Response(serializer.errors,
+                            status=status.HTTP_400_BAD_REQUEST)
+        else:
```
여기도 마찬가지

samjo-nyang commented on sparcs09-api/apps/core/viewsets/item.py:
```diff
@@ -1,53 +1,189 @@
-        pass
+        item = self.get_object()
+
+        if len(Payment.objects.filter(id=item.id)) != 0:
```
여기도 .exists()가 더 좋겠네요

sugoiii committed:
Fixed minor issues

- Item을 get할시 is_deleted일 경우 가져오지 않음.
- 이외 언급된 문제들 수정
